feat: Add ability to create an ingress

cdrage opened the pull request:
feat: Add ability to create an ingress

### What does this PR do?

* Add the option to create an ingress when deploying to Kubernetes
* Compatible with 'Contour' ingress with Kind

### Screenshot/screencast of this PR

<!-- Please include a screenshot or a screencast explaining what is doing this PR -->

https://user-images.githubusercontent.com/6422176/232894496-cbaea036-a14c-46c6-bfa3-bacca629a161.mov




### What issues does this PR fix or reference?

<!-- Please include any related issue from Podman Desktop repository (or from another issue tracker).
-->

Fixes https://github.com/containers/podman-desktop/issues/1322

### How to test this PR?

1. Create a `kind` cluster with port 9090 for http
2. Run 'podman run --name hello -d -p 8080:8080 cdrage/helloworld`
3. Press "Deploy to Kubernetes"
4. Select the ingress option
5. `curl localhost:9090` or visit with the browser.

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

cdrage commented:
Ready for review with the exception of me needing to write tests 👍 

Found a few issues as a result of implementing this PR:
* https://github.com/containers/podman-desktop/issues/2145
* https://github.com/containers/podman-desktop/issues/2148

cdrage commented:
Tests added 🎉 

Also added the ability to create an ingress through the createResource function. So ingresses can be called successfully as well as tested properly against.

lstocchi reviewed (changes requested):
I faced an issue related to the name length. 

```
Error: Ingress.extensions "helloworldcdrage-pod" is invalid: spec.defaultBackend.service.port.name: Invalid value: "helloworldcdrage-pod-9000": must be no more than 15 characters
```

- I pulled your image
- Created a container called helloworldcdrage
- Tried to deploy/create the ingress

jeffmaury reviewed (changes requested):
I don't think ingress should be exposed in addition to routes. If you're on OpenShift routes should be created otherwise ingress.
So I would favor changing the flag to something like 'Expose this deployment outside of the cluster' otherwise I think this would be to much technical for lambda users

cdrage commented on packages/main/src/plugin/kubernetes-client.ts:
```diff
@@ -613,6 +626,13 @@ export class KubernetesClient {
         }
+      } else if (groupVersion.group === 'networking.k8s.io') {
+        // Add networking object support (Ingress for now)
+        const k8sNetworkingApi = this.kubeConfig.makeApiClient(NetworkingV1Api);
```
This was added to so we can test it as well as use it when supplying your own yaml to be pushed to a Kubernetes cluster.

cdrage commented:
> I don't think ingress should be exposed in addition to routes. If you're on OpenShift routes should be created otherwise ingress. So I would favor changing the flag to something like 'Expose this deployment outside of the cluster' otherwise I think this would be to much technical for lambda users

I'll add it so that you can deploy either Ingress or Routes, but not both. OpenShift users shouldn't be getting that option.

I'm a bit confused on changing the flag to "Explore this deployment outside the cluster". We already use "Service" terminology, so should we change that too?

Maybe we can do: "Expose this deployment to access locally (using Kubernetes Ingress)" if that makes sense?

cdrage commented:
> I faced an issue related to the name length.
> 
> ```
> Error: Ingress.extensions "helloworldcdrage-pod" is invalid: spec.defaultBackend.service.port.name: Invalid value: "helloworldcdrage-pod-9000": must be no more than 15 characters
> ```
> 
> * I pulled your image
> * Created a container called helloworldcdrage
> * Tried to deploy/create the ingress

Thanks! It's because I used name mapping instead of port mapping, and the name length is only 15 characters which is very low.

I've updated the PR with the changes. Feel free to try again

jeffmaury commented:
Works on Sandbox but failed on Kind cluster:

![image](https://user-images.githubusercontent.com/695993/233575364-d0ddc81f-7187-4526-8ba4-544a6a983cb1.png)

I tried to deploy centos/httpd-24-centos7:2.4 and selected the 8080 port.

I still don't understand why we don't expose all ports in the Ingress case and not in the Route case

lstocchi commented:
> Works on Sandbox but failed on Kind cluster:

I can face the same error as Jeff

cdrage commented:
> Works on Sandbox but failed on Kind cluster:
> 
> ![image](https://user-images.githubusercontent.com/695993/233575364-d0ddc81f-7187-4526-8ba4-544a6a983cb1.png)
> 
> I tried to deploy centos/httpd-24-centos7:2.4 and selected the 8080 port.
> 
> I still don't understand why we don't expose all ports in the Ingress case and not in the Route case

It works a bit differently in Kubernetes.

It has to be only one because if we were to deploy multiple Ingresses, the "oldest" rule would be the one that the traffic is routed to. 

See: https://kubernetes.github.io/ingress-nginx/how-it-works/ (applies the same to contour)

as well as: https://stackoverflow.com/questions/60501553/kubernetes-multiple-ingress-objects-with-same-configs

All of our ingresses route to the same location `localhost/` since we're using `Contour` Ingress as well as within a local development environment (`kind`).

cdrage commented:
Updated!

Tested with both pod, container and on sandbox + kind.

I used this as an example:
```
podman run -d -p 8080:80 kennethreitz/httpbin

# Second example
podman run -d -p 8081:80 -p 8082:8082 kennethreitz/httpbin
```

lstocchi reviewed (changes requested):
Everything works fine for me.

Only an issue when no port is selected.
If i check the ingress creation checkbox but do not select any port and click deploy i see this error and the button disappear. 
![image](https://user-images.githubusercontent.com/49404737/233690367-aa75f739-8582-4f12-8a3f-d1ca6ee99bbf.png)

Even if i select a port, uncheck/re-check the checkbox, the button never get shown again.

cdrage commented:
> Everything works fine for me.
> 
> 
> 
> Only an issue when no port is selected.
> 
> If i check the ingress creation checkbox but do not select any port and click deploy i see this error and the button disappear. 
> 
> ![image](https://user-images.githubusercontent.com/49404737/233690367-aa75f739-8582-4f12-8a3f-d1ca6ee99bbf.png)
> 
> 
> 
> Even if i select a port, uncheck/re-check the ingress checkbox, the button never get shown again.

Ah! UI validation error. I'll fix that.

cdrage commented:
> Everything works fine for me.
> 
> Only an issue when no port is selected. If i check the ingress creation checkbox but do not select any port and click deploy i see this error and the button disappear. ![image](https://user-images.githubusercontent.com/49404737/233690367-aa75f739-8582-4f12-8a3f-d1ca6ee99bbf.png)
> 
> Even if i select a port, uncheck/re-check the ingress checkbox, the button never get shown again.

Ready for another review!

I added some changes:

- Ability to show a warning
- Show ports correctly (it was trying to show containerPort rather than hostPort which is what we're suppose to ask the user)
- Be able to hit deploy button again
- Only show ingress option when services are enabled (unable to deploy an ingress anyways if a service isn't deployed).
- Only modify bodyPod if the deployment was successful. Bodypod is the "final" kube yaml that is deployed.

cdrage commented:
Opened up this issue that I encountered when implementing: https://github.com/containers/podman-desktop/issues/2201

cdrage commented:
> > > 
> > 
> > 
> > To answer your questions:
> > Single service / port is a limitation of our setup (Kind within a VM on Podman while running Contour).
> > As of right now, if you were to add a common scenario such as providing a host name to an ingress:
> > ```
> > spec:
> >   rules:
> >     - host: hello.world.com
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > To our setup, it would not work.
> > To solve this, we'll have to implement nginx-ingress (more common + more stable than contour issues we've been getting) as well as give stepped instructions.
> > ```
> > 1. Use `nginx-ingress`
> > 
> > 2. `podman container inspect kind-control-plane \   --format '{{ .NetworkSettings.Networks.kind.IPAddress }}'` = Get the IP address
> > 
> > 3. Add `172.18.0.2 hello.world.com` to /etc/hosts
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then you can `curl` the address of `curl hello.world.com`
> > You can also visit it by doing the `--add-host` setting within `podman` to curl the URL:
> > ```
> > docker run \
> >   --add-host hello.world.com:172.18.0.2 \
> >   --net kind \
> >   --rm \
> >   curlimages/curl:7.71.0 hello.world.com
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Right now the best "dev" solution is a single service / host, as the ingress controller is running on the default port of kind (for our example, it's `localhost:9090`).
> > This is effective for development scenarios, but for more typical scenarios outlined above, it'll be a different implementation.
> > We can always have this within the "Deploy to Kubernetes" form, say.. two options such as: "Deploy locally on a single service", and "Deploy to a domain name" with steps on how to setup /etc/hosts (or just ping / port forward via a podman container).
> > Host routing would not work as we'll have to make sure that the `kind` ingress running within the VM is exposable to the web.
> 
> I've been able to deploy 2 services on a Kind cluster using Contour ingress controller using host routing and nip.io: the only issue is how to expose the URLs to the user

You're exposing them with rules? Can you outline the steps. Usually you can't do this unless you have a path rule for Ingress for the same ingress controller location (example, the official kind example shows a `/foo` and `/bar` path: https://kind.sigs.k8s.io/docs/user/ingress/#using-ingress

PR is ready for another review! @lstocchi @jeffmaury

jeffmaury commented:
> > > > 
> > > 
> > > 
> > > To answer your questions:
> > > Single service / port is a limitation of our setup (Kind within a VM on Podman while running Contour).
> > > As of right now, if you were to add a common scenario such as providing a host name to an ingress:
> > > ```
> > > spec:
> > >   rules:
> > >     - host: hello.world.com
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > To our setup, it would not work.
> > > To solve this, we'll have to implement nginx-ingress (more common + more stable than contour issues we've been getting) as well as give stepped instructions.
> > > ```
> > > 1. Use `nginx-ingress`
> > > 
> > > 2. `podman container inspect kind-control-plane \   --format '{{ .NetworkSettings.Networks.kind.IPAddress }}'` = Get the IP address
> > > 
> > > 3. Add `172.18.0.2 hello.world.com` to /etc/hosts
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Then you can `curl` the address of `curl hello.world.com`
> > > You can also visit it by doing the `--add-host` setting within `podman` to curl the URL:
> > > ```
> > > docker run \
> > >   --add-host hello.world.com:172.18.0.2 \
> > >   --net kind \
> > >   --rm \
> > >   curlimages/curl:7.71.0 hello.world.com
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Right now the best "dev" solution is a single service / host, as the ingress controller is running on the default port of kind (for our example, it's `localhost:9090`).
> > > This is effective for development scenarios, but for more typical scenarios outlined above, it'll be a different implementation.
> > > We can always have this within the "Deploy to Kubernetes" form, say.. two options such as: "Deploy locally on a single service", and "Deploy to a domain name" with steps on how to setup /etc/hosts (or just ping / port forward via a podman container).
> > > Host routing would not work as we'll have to make sure that the `kind` ingress running within the VM is exposable to the web.
> > 
> > 
> > I've been able to deploy 2 services on a Kind cluster using Contour ingress controller using host routing and nip.io: the only issue is how to expose the URLs to the user
> 
> You're exposing them with rules? Can you outline the steps. Usually you can't do this unless you have a path rule for Ingress for the same ingress controller location (example, the official kind example shows a `/foo` and `/bar` path: https://kind.sigs.k8s.io/docs/user/ingress/#using-ingress
> 
> PR is ready for another review! @lstocchi @jeffmaury

Please find the 2 deployments:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: httpbin1
  name: httpbin1
spec:
  replicas: 1
  selector:
    matchLabels:
      app: httpbin1
  template:
    metadata:
      labels:
        app: httpbin1
    spec:
      containers:
      - image: docker.io/kennethreitz/httpbin
        name: httpbin1
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: httpbin1
  name: httpbin1
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: httpbin1
  sessionAffinity: None
  type: ClusterIP
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: httpbin1
  labels:
    app: httpbin1
spec:
  rules:
  - host: httpbin1.127.0.0.1.nip.io
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: httpbin1
            port:
              number: 80

```

```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: httpbin2
  name: httpbin2
spec:
  replicas: 1
  selector:
    matchLabels:
      app: httpbin2
  template:
    metadata:
      labels:
        app: httpbin2
    spec:
      containers:
      - image: docker.io/kennethreitz/httpbin
        name: httpbin2
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: httpbin2
  name: httpbin2
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: httpbin2
  sessionAffinity: None
  type: ClusterIP
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: httpbin2
  labels:
    app: httpbin2
spec:
  rules:
  - host: httpbin2.127.0.0.1.nip.io
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: httpbin2
            port:
              number: 80

```

cdrage commented:
> > > > > 
> > > > 
> > > > 
> > > > To answer your questions:
> > > > Single service / port is a limitation of our setup (Kind within a VM on Podman while running Contour).
> > > > As of right now, if you were to add a common scenario such as providing a host name to an ingress:
> > > > ```
> > > > spec:
> > > >   rules:
> > > >     - host: hello.world.com
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > To our setup, it would not work.
> > > > To solve this, we'll have to implement nginx-ingress (more common + more stable than contour issues we've been getting) as well as give stepped instructions.
> > > > ```
> > > > 1. Use `nginx-ingress`
> > > > 
> > > > 2. `podman container inspect kind-control-plane \   --format '{{ .NetworkSettings.Networks.kind.IPAddress }}'` = Get the IP address
> > > > 
> > > > 3. Add `172.18.0.2 hello.world.com` to /etc/hosts
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Then you can `curl` the address of `curl hello.world.com`
> > > > You can also visit it by doing the `--add-host` setting within `podman` to curl the URL:
> > > > ```
> > > > docker run \
> > > >   --add-host hello.world.com:172.18.0.2 \
> > > >   --net kind \
> > > >   --rm \
> > > >   curlimages/curl:7.71.0 hello.world.com
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Right now the best "dev" solution is a single service / host, as the ingress controller is running on the default port of kind (for our example, it's `localhost:9090`).
> > > > This is effective for development scenarios, but for more typical scenarios outlined above, it'll be a different implementation.
> > > > We can always have this within the "Deploy to Kubernetes" form, say.. two options such as: "Deploy locally on a single service", and "Deploy to a domain name" with steps on how to setup /etc/hosts (or just ping / port forward via a podman container).
> > > > Host routing would not work as we'll have to make sure that the `kind` ingress running within the VM is exposable to the web.
> > > 
> > > 
> > > I've been able to deploy 2 services on a Kind cluster using Contour ingress controller using host routing and nip.io: the only issue is how to expose the URLs to the user
> > 
> > 
> > You're exposing them with rules? Can you outline the steps. Usually you can't do this unless you have a path rule for Ingress for the same ingress controller location (example, the official kind example shows a `/foo` and `/bar` path: https://kind.sigs.k8s.io/docs/user/ingress/#using-ingress
> > PR is ready for another review! @lstocchi @jeffmaury
> 
> Please find the 2 deployments:
> 
> ```
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   labels:
>     app: httpbin1
>   name: httpbin1
> spec:
>   replicas: 1
>   selector:
>     matchLabels:
>       app: httpbin1
>   template:
>     metadata:
>       labels:
>         app: httpbin1
>     spec:
>       containers:
>       - image: docker.io/kennethreitz/httpbin
>         name: httpbin1
> ---
> apiVersion: v1
> kind: Service
> metadata:
>   labels:
>     app: httpbin1
>   name: httpbin1
> spec:
>   ports:
>   - port: 80
>     protocol: TCP
>     targetPort: 80
>   selector:
>     app: httpbin1
>   sessionAffinity: None
>   type: ClusterIP
> ---
> apiVersion: networking.k8s.io/v1
> kind: Ingress
> metadata:
>   name: httpbin1
>   labels:
>     app: httpbin1
> spec:
>   rules:
>   - host: httpbin1.127.0.0.1.nip.io
>     http:
>       paths:
>       - path: /
>         pathType: Prefix
>         backend:
>           service:
>             name: httpbin1
>             port:
>               number: 80
> ```
> 
> ```
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   labels:
>     app: httpbin2
>   name: httpbin2
> spec:
>   replicas: 1
>   selector:
>     matchLabels:
>       app: httpbin2
>   template:
>     metadata:
>       labels:
>         app: httpbin2
>     spec:
>       containers:
>       - image: docker.io/kennethreitz/httpbin
>         name: httpbin2
> ---
> apiVersion: v1
> kind: Service
> metadata:
>   labels:
>     app: httpbin2
>   name: httpbin2
> spec:
>   ports:
>   - port: 80
>     protocol: TCP
>     targetPort: 80
>   selector:
>     app: httpbin2
>   sessionAffinity: None
>   type: ClusterIP
> ---
> apiVersion: networking.k8s.io/v1
> kind: Ingress
> metadata:
>   name: httpbin2
>   labels:
>     app: httpbin2
> spec:
>   rules:
>   - host: httpbin2.127.0.0.1.nip.io
>     http:
>       paths:
>       - path: /
>         pathType: Prefix
>         backend:
>           service:
>             name: httpbin2
>             port:
>               number: 80
> ```

Yes, unfortunately that's different than our Ingress example for httpbin which is:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: httpbin
  labels:
    app: httpbin
spec:
  defaultBackend:
    service:
      name: httpbin
      port:
        number: 80
```

Which uses `defaultBackend` and no rules / specifying domain name. 

When we specify the rules / domain name, we are no longer able to access it locally unless we edit `/etc/hosts` or use `nip.io` which the new-to-kubernetes developer may not be familiar with. So that's why we were implementing this "simple" way.

I've been basing this PR on Contour as it's the only ingress we support, the examples use a defaultBackend / 1 service, so that we can easily visit it at localhost:9090.

In a follow up PR we should add the functionality of supplying a domain name as well as documentation on how to use `nip.io` / how it works. We would have to add documentation to the website as well.

jeffmaury commented:
OK let's delay the nio.io based solution. 

I tried it but httpbin but it failed for me:

![image](https://user-images.githubusercontent.com/695993/234561245-40228ae2-a013-4c93-9cf1-ae0c16c42123.png)

cdrage commented:
> OK let's delay the nio.io based solution.
> 
> I tried it but httpbin but it failed for me:
> 
> ![image](https://user-images.githubusercontent.com/695993/234561245-40228ae2-a013-4c93-9cf1-ae0c16c42123.png)

Updated. I had to rebase, but it is working now with port selection.

@lstocchi can you test too?

lstocchi commented:
> @lstocchi can you test too?

Tried again @cdrage i still see the issue with the button not being enabled after not selecting a port
![ingress-creation-issue](https://user-images.githubusercontent.com/49404737/234600365-e3f4ba57-88a3-4fee-89fd-1e2ffd2651ae.gif)

lstocchi commented:
> > 
> 
> ` deployStarted = false;` got removed when I rebased.
> 
> Can you try again?

Thanks, that works fine now 👍 
I faced this though... any idea what's this?
![image](https://user-images.githubusercontent.com/49404737/234624244-32f59715-a01c-4775-9214-0c37bbe565a5.png)

cdrage commented:
> > > 
> > 
> > 
> > ` deployStarted = false;` got removed when I rebased.
> > Can you try again?
> 
> Thanks, that works fine now 👍 I faced this though... any idea what's this? ![image](https://user-images.githubusercontent.com/49404737/234624244-32f59715-a01c-4775-9214-0c37bbe565a5.png)

Unable to replicate :( I also tested against minikube as per your screenshot. I just rebased (again), are you able to try again / show a video from creating a container to the error?

lstocchi reviewed (approved):
Tried today with a new, clean  kind cluster and i don't face that issue anymore so i guess it was broken on my machine. 
For what i can see, it works fine. LGTM

themr0c commented on packages/renderer/src/lib/pod/DeployPodToKube.svelte:
```diff
@@ -260,6 +355,45 @@ function updateKubeResult() {
+            class=""
+            required />
+          <span class="text-gray-300 text-sm ml-1"
+            >Create a Kubernetes ingress to get access to the exposed ports of this pod. An ingress controller is
```
Consider rephrasing:

Create a Kubernetes ingress to get access to the ports that this pod exposes, at the default ingress controller location. Example: on a default Kind cluster created with Podman Desktop: `http://localhost:9090`. Requirements: your cluster has an ingress controller.

cdrage replied:
Updated! :)

jeffmaury commented:
LGTM.

cdrage commented:
Enabled auto-merge & rebased 🎉
